spring-cloud-stream-binder-kinesis-core/src/main/java/org/springframework/cloud/stream/binder/kinesis/provisioning/KinesisStreamProvisioner.java

```diff
@@ -55,19 +61,40 @@ public KinesisStreamProvisioner(AmazonKinesis amazonKinesis,
 	@Override
 	public ProducerDestination provisionProducerDestination(String name,
 			ExtendedProducerProperties<KinesisProducerProperties> properties) throws ProvisioningException {
+		logger.info("Using Kinesis stream for outbound: " + name);
 
-		KinesisProducerDestination producer = new KinesisProducerDestination(name);
+		KinesisStream stream = createOrUpdate(name, properties.getPartitionCount());
 
-		return producer;
+		return new KinesisProducerDestination(stream.getName(), stream.getShards());
 	}
 
 	@Override
 	public ConsumerDestination provisionConsumerDestination(String name, String group,
 			ExtendedConsumerProperties<KinesisConsumerProperties> properties) throws ProvisioningException {
+		logger.info("Using Kinesis stream for inbound: " + name);
+
+		int shardCount = properties.getInstanceCount() * properties.getConcurrency();
+
+		KinesisStream stream = createOrUpdate(name, shardCount);
+
+		return new KinesisConsumerDestination(stream.getName(), stream.getShards());
+	}
+
+	private KinesisStream createOrUpdate(String name, Integer shards) {
+		try {
+			DescribeStreamResult streamResult = amazonKinesis.describeStream(name);
+			logger.info("Stream found, using existing stream");
 
-		KinesisConsumerDestination consumer = new KinesisConsumerDestination(name);
+			return new KinesisStream(name, streamResult.getStreamDescription().getShards().size());
```

Comment:
No problem, I'm taking a narrow approach for this contribution because I'm just learning both Spring Cloud Stream and Kinesis and don't want the PR to be open for too long. Thank you for your patience! :-)